diff -urN atompaw-3.0.old/src/aeatom.f90 atompaw-3.0.new/src/aeatom.f90
--- atompaw-3.0.old/src/aeatom.f90	2011-01-29 23:53:52.000000000 +0100
+++ atompaw-3.0.new/src/aeatom.f90	2011-04-02 08:15:52.000000000 +0100
@@ -427,6 +427,23 @@
              ENDDO
           ENDIF
        ENDDO
+
+       ! Added by MS Martin Stankovski
+       ! Output all the wavefunctions for easy plotting in gnuplot
+       OPEN(13,file=trim(AEpot%sym)//'_all_wfn.dat')
+       WRITE(13,'(3(a),e15.7)') '# Atom: ',AEpot%sym,' Z: ',nz
+       WRITE(13,'(a)') '#  n  l     occupancy       energy'
+       DO io=1,norbit
+         WRITE(13,'(a,i2,1x,i2,4x,1p,2e15.7)') '# ',&
+               AEOrbit%np(io),AEOrbit%l(io),&
+               AEOrbit%occ(io),AEOrbit%eig(io)
+       ENDDO
+       DO i=1,n
+          WRITE(13,'(100(1p,4e25.17))') AEGrid%r(i),(AEOrbit%wfn(i,io), io=1,norbit)
+       ENDDO
+       CLOSE(13)
+       ! end of modification by Martin Stankovski
+
        qcal=integrator(AEGrid,denout)
        WRITE(6,*) 'qcal electrons = ',qcal, electrons
        !  rescale density
diff -urN atompaw-3.0.old/src/atompaw_prog.f90 atompaw-3.0.new/src/atompaw_prog.f90
--- atompaw-3.0.old/src/atompaw_prog.f90	2011-01-29 23:53:52.000000000 +0100
+++ atompaw-3.0.new/src/atompaw_prog.f90	2011-04-02 08:15:52.000000000 +0100
@@ -83,6 +83,9 @@
   REAL(8) :: tildekin,tildepot,onehat,onehartree,ctexc,ctexc1,cexc1,texc,vexc1
   REAL(8) :: vtexc,vtexc1,oneenergy,ekin,vlocal,tq,fac,stuff,term,sqr4pi
   REAL(8) :: ctctse,cthatse,selfenergy
+  ! start modification by Martin Stankovski (MS)
+  REAL(8) :: first_point
+  ! end modification by Martin Stankovski 
   REAL(8), ALLOCATABLE :: ttphi(:),soij(:),stij(:),svij(:)
   REAL(8), ALLOCATABLE :: shij(:),snij(:),dum(:),dum1(:),rh(:),rth(:)
   REAL(8), ALLOCATABLE :: shartree(:),sshartree(:,:,:),wf(:),twf(:)
@@ -461,6 +464,57 @@
   ENDDO
   CLOSE(ifout)
 
+  ! start modification by Martin Stankovski (MS)
+  ! MS  - output density*r^2 for abinit calc. 
+  OPEN(ifout,file=trim(AEpot%sym)//'_full_density.dat')
+  WRITE(ifout,'(3(a),e15.7)') '# Atom: ',AEpot%sym,' Z: ',AEpot%nz
+  ! Interpolate first point (Lagrange three-point interp.)
+  first_point = &
+  &   ((PAW%den(2)+FC%coreden(2))/(r(2)*r(2)))*r(3)*r(4) &
+  &          /((r(2)-r(3))*(r(2)-r(4))) &
+  & + ((PAW%den(3)+FC%coreden(3))/(r(3)*r(3)))*r(2)*r(4) &
+  &          /((r(3)-r(2))*(r(3)-r(4))) &
+  & + ((PAW%den(4)+FC%coreden(4))/(r(4)*r(4)))*r(2)*r(3) &
+  &          /((r(4)-r(2))*(r(4)-r(3)))
+  WRITE(ifout,'(2(1p,4e25.17))') 0.0d0,first_point/(4.d0*pi)
+  DO i=2,n
+    IF ((PAW%den(i)+FC%coreden(i))<machine_zero) EXIT
+    WRITE(ifout,'(2(1p,4e25.17))') r(i), &
+    &       ((PAW%den(i)+FC%coreden(i))/(r(i)*r(i)*4.d0*pi))
+  END DO
+  CLOSE(ifout)
+
+    ! MS - output density*r^2 for abinit calc. 
+  OPEN(ifout,file=trim(AEpot%sym)//'_valence_density.dat')
+  WRITE(ifout,'(3(a),e15.7)') '# Atom: ',AEpot%sym,' Z: ',AEpot%nz
+  ! Interpolate first point (Lagrange three-point interp.)
+  first_point = &
+  &   (PAW%den(2)/(r(2)*r(2)))*r(3)*r(4)/((r(2)-r(3))*(r(2)-r(4))) &
+  & + (PAW%den(3)/(r(3)*r(3)))*r(2)*r(4)/((r(3)-r(2))*(r(3)-r(4))) &
+  & + (PAW%den(4)/(r(4)*r(4)))*r(2)*r(3)/((r(4)-r(2))*(r(4)-r(3)))
+  WRITE(ifout,'(2(1p,4e25.17))') 0.0d0,first_point/(4.d0*pi)
+  DO i=2,n
+    IF (PAW%den(i)<machine_zero) EXIT
+    WRITE(ifout,'(2(1p,4e25.17))') r(i),(PAW%den(i)/(r(i)*r(i)*4.d0*pi))
+  END DO
+  CLOSE(ifout)
+
+  ! MS - output core density*r^2 for abinit calc. 
+  OPEN(ifout,file=trim(AEpot%sym)//'_core_density.dat')
+  WRITE(ifout,'(3(a),e15.7)') '# Atom: ',AEpot%sym,' Z: ',AEpot%nz
+  ! Interpolate first point (Lagrange three-point interp.)
+  first_point = &
+  &   (FC%coreden(2)/(r(2)*r(2)))*r(3)*r(4)/((r(2)-r(3))*(r(2)-r(4))) &
+  & + (FC%coreden(3)/(r(3)*r(3)))*r(2)*r(4)/((r(3)-r(2))*(r(3)-r(4))) &
+  & + (FC%coreden(4)/(r(4)*r(4)))*r(2)*r(3)/((r(4)-r(2))*(r(4)-r(3)))
+  WRITE(ifout,'(2(1p,4e25.17))') 0.0d0,first_point/(4.d0*pi)
+  DO i=2,n
+    IF (FC%coreden(i)<machine_zero) EXIT
+    WRITE(ifout,'(2(1p,4e25.17))') r(i),(FC%coreden(i)/(r(i)*r(i)*4.d0*pi))
+  END DO
+  CLOSE(ifout)
+  ! end of modification by MS Martin Stankovski
+
   OPEN(ifout,file='potential', form='formatted')
   DO i=1,n
      IF (ABS(rv(i))<machine_zero) rv(i)=0
