*** bigdft-abi-1.0.4/src/init/gautowav.f90	Mon Jul  9 16:43:33 2012
--- bigdft-patch/src/init/gautowav.f90	Mon Jun 10 11:16:37 2013
***************
*** 675,680 ****
--- 675,681 ----
    integer :: i_stat,i_all,ishell,iexpo,icoeff,iat,isat,ng,l,m,i,nterm,ig
    integer :: nterms_max,nterms,iterm,n_gau,ml1,mu1,ml2,mu2,ml3,mu3 !n(c) iscoeff
    real(gp) :: rx,ry,rz,gau_a
+   real(gp) :: gau_cut !only for PAW
    integer, dimension(nterm_max) :: lx,ly,lz
    real(gp), dimension(nterm_max) :: fac_arr
    real(wp), allocatable, dimension(:,:,:) :: work
***************
*** 746,762 ****
                      gau_a=G%xp(1,iexpo+ig-1)
                      n_gau=lx(i)
                      !print *,'x',gau_a,nterm,ncplx,kx,ky,kz,ml1,mu1,lr%d%n1
!                     call gauss_to_daub_k(hx,kx*hx,ncplx,1,ncplx,fac_arr(i),rx,gau_a,n_gau,&
                           lr%ns1,lr%d%n1,ml1,mu1,&
                           wx(1,0,1,iterm),work,nw,perx) 
                      n_gau=ly(i)
                      !print *,'y',ml2,mu2,lr%d%n2
!                     call gauss_to_daub_k(hy,ky*hy,ncplx,1,ncplx,wfn_gau(icoeff),ry,gau_a,n_gau,&
                           lr%ns2,lr%d%n2,ml2,mu2,&
                           wy(1,0,1,iterm),work,nw,pery) 
                      n_gau=lz(i) 
                      !print *,'z',ml3,mu3,lr%d%n3
!                     call gauss_to_daub_k(hz,kz*hz,ncplx,1,ncplx,G%psiat(:,iexpo+ig-1),rz,gau_a,n_gau,&
                           lr%ns3,lr%d%n3,ml3,mu3,&
                           wz(1,0,1,iterm),work,nw,perz)
                      iterm=iterm+1
--- 747,763 ----
                      gau_a=G%xp(1,iexpo+ig-1)
                      n_gau=lx(i)
                      !print *,'x',gau_a,nterm,ncplx,kx,ky,kz,ml1,mu1,lr%d%n1
!                     call gauss_to_daub_k(hx,kx*hx,ncplx,1,ncplx,fac_arr(i),rx,gau_a,gau_cut,n_gau,&
                           lr%ns1,lr%d%n1,ml1,mu1,&
                           wx(1,0,1,iterm),work,nw,perx) 
                      n_gau=ly(i)
                      !print *,'y',ml2,mu2,lr%d%n2
!                     call gauss_to_daub_k(hy,ky*hy,ncplx,1,ncplx,wfn_gau(icoeff),ry,gau_a,gau_cut,n_gau,&
                           lr%ns2,lr%d%n2,ml2,mu2,&
                           wy(1,0,1,iterm),work,nw,pery) 
                      n_gau=lz(i) 
                      !print *,'z',ml3,mu3,lr%d%n3
!                     call gauss_to_daub_k(hz,kz*hz,ncplx,1,ncplx,G%psiat(:,iexpo+ig-1),rz,gau_a,gau_cut,n_gau,&
                           lr%ns3,lr%d%n3,ml3,mu3,&
                           wz(1,0,1,iterm),work,nw,perz)
                      iterm=iterm+1
