*** bigdft-abi-1.0.4/src/hpsiortho.f90	Wed Jan 30 10:10:56 2013
--- bigdft-patch/src/hpsiortho.f90	Tue Jun 11 09:49:22 2013
***************
*** 720,726 ****
                 if(at%npspcode(1)==7) then
                   call atom_projector_paw(ikpt,iat,0,istart_c,iproj,&
                        nlpspd%nprojel,&
!                       Lzd%Glr,Lzd%hgrids(1),Lzd%hgrids(2),Lzd%hgrids(3),rxyz(1,iat),at,orbs,&
                        nlpspd%plr(iat),proj,nwarnings,proj_G(iatype))
                 else
                   call atom_projector(ikpt,iat,0,istart_c,iproj,&
--- 720,727 ----
                 if(at%npspcode(1)==7) then
                   call atom_projector_paw(ikpt,iat,0,istart_c,iproj,&
                        nlpspd%nprojel,&
!                       Lzd%Glr,Lzd%hgrids(1),Lzd%hgrids(2),Lzd%hgrids(3),&
!                       paw%rpaw(iatype),rxyz(1,iat),at,orbs,&
                        nlpspd%plr(iat),proj,nwarnings,proj_G(iatype))
                 else
                   call atom_projector(ikpt,iat,0,istart_c,iproj,&
