--- bigdft-1.2.0.1.old/src/sumrho.F90	2009-07-01 10:49:36.252793679 +0200
+++ bigdft-1.2.0.1.new/src/sumrho.F90	2009-07-01 10:50:08.693028765 +0200
@@ -92,9 +92,12 @@
                MPI_DOUBLE_PRECISION,MPI_SUM,MPI_COMM_WORLD,ierr)
         end do
      else
-         call MPI_ALLREDUCE(MPI_IN_PLACE,rho_p,lr%d%n1i*lr%d%n2i*lr%d%n3i*nspin,&
-              MPI_DOUBLE_PRECISION,MPI_SUM,MPI_COMM_WORLD,ierr)
-         !stop 'rsflag active in sumrho.f90, check MPI2 implementation'
+#ifdef MPI2
+        call MPI_ALLREDUCE(MPI_IN_PLACE,rho_p,lr%d%n1i*lr%d%n2i*lr%d%n3i*nspin,&
+             MPI_DOUBLE_PRECISION,MPI_SUM,MPI_COMM_WORLD,ierr)
+#else
+        stop 'rsflag active in sumrho.f90, check MPI2 implementation'
+#endif
      end if
      call timing(iproc,'Rho_commun    ','OF')
      call timing(iproc,'Rho_comput    ','ON')
