*** bigdft-abi-1.0.4/src/wfn_opt/applyh.f90	Tue May 28 15:42:28 2013
--- bigdft-patch/src/wfn_opt/applyh.f90	Thu Jun  6 11:29:13 2013
***************
*** 1104,1110 ****
          istart_c=1
          if(at%npspcode(iatype)==7) then
            call atom_projector_paw(ikpt,iat,idir,istart_c,iproj,nlpspd%nprojel,&
!                lr,hx,hy,hz,rxyz(1,iat),at,orbs,nlpspd%plr(iat),proj,nwarnings,proj_G(iatype))
          else
            call atom_projector(ikpt,iat,idir,istart_c,iproj,nlpspd%nprojel,&
                 lr,hx,hy,hz,rxyz(1,iat),at,orbs,nlpspd%plr(iat),proj,nwarnings)
--- 1104,1110 ----
          istart_c=1
          if(at%npspcode(iatype)==7) then
            call atom_projector_paw(ikpt,iat,idir,istart_c,iproj,nlpspd%nprojel,&
!                lr,hx,hy,hz,paw%rpaw(iatype),rxyz(1,iat),at,orbs,nlpspd%plr(iat),proj,nwarnings,proj_G(iatype))
          else
            call atom_projector(ikpt,iat,idir,istart_c,iproj,nlpspd%nprojel,&
                 lr,hx,hy,hz,rxyz(1,iat),at,orbs,nlpspd%plr(iat),proj,nwarnings)
***************
*** 1116,1122 ****
             istart_c=1
             if(at%npspcode(iatype)==7) then
             !    PAW case:
!               call apply_atproj_iorb_paw(iat,iorb,nlpspd%nprojel,&
                     at,orbs,wfd,nlpspd%plr(iat),proj,&
                     psi(ispsi),hpsi(ispsi),eproj_sum,proj_G(iatype),paw)
             else
--- 1116,1123 ----
             istart_c=1
             if(at%npspcode(iatype)==7) then
             !    PAW case:
!               call apply_atproj_iorb_paw(iat,iorb,ispsi,istart_c,&
!                    nlpspd%nprojel,&
                     at,orbs,wfd,nlpspd%plr(iat),proj,&
                     psi(ispsi),hpsi(ispsi),eproj_sum,proj_G(iatype),paw)
             else
