diff -urN bigdft-abi-1.0.4.old/configure bigdft-abi-1.0.4.new/configure
--- bigdft-abi-1.0.4.old/configure	2012-10-05 15:10:27.000000000 +0200
+++ bigdft-abi-1.0.4.new/configure	2013-06-11 16:51:00.000000000 +0200
@@ -9185,15 +9185,15 @@
   withlibabinit=no
 fi
 
-    as_ac_File=`$as_echo "ac_cv_file_$ac_libabinit_dir/include/ab6_moldyn.$ax_fc_mod_ext" | $as_tr_sh`
-{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for $ac_libabinit_dir/include/ab6_moldyn.$ax_fc_mod_ext" >&5
-$as_echo_n "checking for $ac_libabinit_dir/include/ab6_moldyn.$ax_fc_mod_ext... " >&6; }
+    as_ac_File=`$as_echo "ac_cv_file_$ac_libabinit_dir/include/ab7_moldyn.$ax_fc_mod_ext" | $as_tr_sh`
+{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for $ac_libabinit_dir/include/ab7_moldyn.$ax_fc_mod_ext" >&5
+$as_echo_n "checking for $ac_libabinit_dir/include/ab7_moldyn.$ax_fc_mod_ext... " >&6; }
 if eval \${$as_ac_File+:} false; then :
   $as_echo_n "(cached) " >&6
 else
   test "$cross_compiling" = yes &&
   as_fn_error $? "cannot check for file existence when cross compiling" "$LINENO" 5
-if test -r "$ac_libabinit_dir/include/ab6_moldyn.$ax_fc_mod_ext"; then
+if test -r "$ac_libabinit_dir/include/ab7_moldyn.$ax_fc_mod_ext"; then
   eval "$as_ac_File=yes"
 else
   eval "$as_ac_File=no"
@@ -9208,15 +9208,15 @@
   moldyn="no"
 fi
 
-    as_ac_File=`$as_echo "ac_cv_file_$ac_libabinit_dir/include/ab6_symmetry.$ax_fc_mod_ext" | $as_tr_sh`
-{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for $ac_libabinit_dir/include/ab6_symmetry.$ax_fc_mod_ext" >&5
-$as_echo_n "checking for $ac_libabinit_dir/include/ab6_symmetry.$ax_fc_mod_ext... " >&6; }
+    as_ac_File=`$as_echo "ac_cv_file_$ac_libabinit_dir/include/ab7_symmetry.$ax_fc_mod_ext" | $as_tr_sh`
+{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for $ac_libabinit_dir/include/ab7_symmetry.$ax_fc_mod_ext" >&5
+$as_echo_n "checking for $ac_libabinit_dir/include/ab7_symmetry.$ax_fc_mod_ext... " >&6; }
 if eval \${$as_ac_File+:} false; then :
   $as_echo_n "(cached) " >&6
 else
   test "$cross_compiling" = yes &&
   as_fn_error $? "cannot check for file existence when cross compiling" "$LINENO" 5
-if test -r "$ac_libabinit_dir/include/ab6_symmetry.$ax_fc_mod_ext"; then
+if test -r "$ac_libabinit_dir/include/ab7_symmetry.$ax_fc_mod_ext"; then
   eval "$as_ac_File=yes"
 else
   eval "$as_ac_File=no"
