diff -urN bigdft-1.2.0.1/src/geometry.f90 ../bigdft-1.2.0.1/src/geometry.f90
--- bigdft-1.2.0.1/src/geometry.f90	2008-12-09 17:41:41.000000000 +0100
+++ bigdft-1.2.0.1/src/geometry.f90	2009-05-26 18:51:18.620210000 +0200
@@ -45,7 +45,7 @@
   !write the first position
   if (iproc.eq.0) call  wtposout(ncount_cluster,etot,rxyz,at)
   !    Open a log file for conjgrad
-  open(unit=16,file='conjgrad.prc',access='append')
+  open(unit=16,file='conjgrad.prc',position='append')
 
   if (in%betax <= 0._gp) then
      call detbetax(nproc,iproc,at,rxyz,rst,in)
