--- bigdft-1.2.0.1.old/src/PSolver/base.f90	2009-06-01 20:40:35.000000000 +0200
+++ bigdft-1.2.0.1.new/src/PSolver/base.f90	2009-06-01 20:41:31.000000000 +0200
@@ -31,6 +31,8 @@
   integer, parameter :: mpidtypg=MPI_DOUBLE_PRECISION
   !integer, parameter :: mpidtypw=MPI_REAL,mpidtypd=MPI_REAL !in case of single precision
 
+  real(kind=4), parameter :: memorylimit=3.e0
+
   !flag for GPU computing, if CUDA libraries are present
   !the parameter flag is removed to allow each processor to modify it following the needs
   !in that case if a GPU is present a given MPI processor may or not perform a GPU calculation
