diff -urN libxc-1.1.0.old/src/gga_x_b88.c libxc-1.1.0.new/src/gga_x_b88.c
--- libxc-1.1.0.old/src/gga_x_b88.c	2011-12-28 19:27:38.000000000 +0100
+++ libxc-1.1.0.new/src/gga_x_b88.c	2011-12-28 19:19:12.000000000 +0100
@@ -121,7 +121,7 @@
   f2 = 1.0 + gamma*beta*x*asinh(x);
   *f = 1.0 + f1/f2;
 
-  if(p->func == 5) /* k_thakkar */
+  /* if(p->func == 5) */ /* k_thakkar */
   df1 = 2.0*beta/X_FACTOR_C*x;
   df2 = gamma*beta*(asinh(x) + x/SQRT(1.0 + x*x));
 
