diff -urN wannier90-1.2.old/src/disentangle.F90 wannier90-1.2.new/src/disentangle.F90
--- wannier90-1.2.old/src/disentangle.F90	2009-11-28 20:09:26.000000000 +0100
+++ wannier90-1.2.new/src/disentangle.F90	2010-04-15 18:01:19.000000000 +0200
@@ -1838,9 +1838,9 @@
       if (ierr/=0) call io_error('Error allocating cham in dis_extract')
 
       if (.not.dis_converged) then
-         write(stdout,'(/5x,a)') '<<< Warning: Maximum number of disentanglement &
-              &iterations reached >>>'
-         write(stdout,'(10x,a)')  '<<< Disentanglement convergence criteria not satisfied >>>'
+         write(stdout,'(/5x,a)') &
+          '<<< Warning: Maximum number of disentanglement iterations reached >>>'
+         write(stdout,'(10x,a)') '<<< Disentanglement convergence criteria not satisfied >>>'
       endif
 
       if(index(devel_flag,'compspace')>0) then
@@ -2622,8 +2622,8 @@
 
 
       if (.not.dis_converged) then
-         write(stdout,'(/5x,a)') '<<< Warning: Maximum number of disentanglement &
-              &iterations reached >>>'
+         write(stdout,'(/5x,a)') &
+           '<<< Warning: Maximum number of disentanglement iterations reached >>>'
          write(stdout,'(10x,a)')  '<<< Disentanglement convergence criteria not satisfied >>>'
       endif
 
diff -urN wannier90-1.2.old/src/overlap.F90 wannier90-1.2.new/src/overlap.F90
--- wannier90-1.2.old/src/overlap.F90	2009-11-22 02:00:04.000000000 +0100
+++ wannier90-1.2.new/src/overlap.F90	2010-04-15 18:01:19.000000000 +0200
@@ -142,8 +142,8 @@
                    nn_found=.true.
                    nn=inn
                 else
-                   call io_error('Error reading '//trim(seedname)//'.mmn.&
-                        & More than one matching nearest neighbour found')
+                   call io_error('Error reading '//trim(seedname)// &
+                        '.mmn. More than one matching nearest neighbour found')
                 endif
              endif
           end do
diff -urN wannier90-1.2.old/src/parameters.F90 wannier90-1.2.new/src/parameters.F90
--- wannier90-1.2.old/src/parameters.F90	2010-01-26 10:40:44.000000000 +0100
+++ wannier90-1.2.new/src/parameters.F90	2010-04-15 18:01:19.000000000 +0200
@@ -1090,8 +1090,9 @@
        write(stdout,'(1x,a)') '+----------------------------------------------------------------------------+'
        do nsp=1,num_species
           do nat=1,atoms_species_num(nsp)
-             write(stdout,'(1x,a1,1x,a2,1x,i3,3F10.5,3x,a1,1x,3F10.5,4x,a1)') '|',atoms_symbol(nsp),nat,atoms_pos_frac(:,nat,nsp),&
-                  '|',atoms_pos_cart(:,nat,nsp)*lenconfac,'|'
+             write(stdout,'(1x,a1,1x,a2,1x,i3,3F10.5,3x,a1,1x,3F10.5,4x,a1)') &
+&                 '|',atoms_symbol(nsp),nat,atoms_pos_frac(:,nat,nsp),&
+&                 '|',atoms_pos_cart(:,nat,nsp)*lenconfac,'|'
           end do
        end do
        write(stdout,'(1x,a)') '*----------------------------------------------------------------------------*'
@@ -1109,8 +1110,8 @@
        write(stdout,'(1x,a)') '|     Frac. Coord.   l mr  r        z-axis               x-axis          Z/a |'
        write(stdout,'(1x,a)') '+----------------------------------------------------------------------------+'
        do nsp=1,num_proj
-          write(stdout,'(1x,a1,3(1x,f5.2),1x,i2,1x,i2,1x,i2,3(1x,f6.3),3(1x,f6.3),&
-               & 2x,f4.1,1x,a1)')  '|',proj_site(1,nsp),proj_site(2,nsp),&
+          write(stdout,'(1x,a1,3(1x,f5.2),1x,i2,1x,i2,1x,i2,3(1x,f6.3),3(1x,f6.3),2x,f4.1,1x,a1)')&
+&              '|',proj_site(1,nsp),proj_site(2,nsp),&
                proj_site(3,nsp),proj_l(nsp), proj_m(nsp),proj_radial(nsp),&
                proj_z(1,nsp),proj_z(2,nsp),proj_z(3,nsp),proj_x(1,nsp),&
                proj_x(2,nsp),proj_x(3,nsp),proj_zona(nsp),'|'
@@ -2806,8 +2807,8 @@
            dummy=utility_strip(in_data(line))
            dummy=adjustl(dummy)
            pos1=index(dummy,':')
-           if(pos1==0) call io_error('param_read_projection: malformed projection &
-                &definition: '//trim(dummy))
+           if(pos1==0) &
+              call io_error('param_read_projection: malformed projection definition: '//trim(dummy))
            sites=0
            ctemp=dummy(:pos1-1)
            ! Read the atomic site
@@ -3048,11 +3049,11 @@
               endif
            end if
            if(sites==-1) then
-              if(counter+sum(ang_states) > num_proj) call io_error('param_get_projection: &
-                   &too many projections defined')
+              if(counter+sum(ang_states) > num_proj) &
+                 call io_error('param_get_projection: too many projections defined')
            else
-              if(counter+sites*sum(ang_states) > num_proj) call io_error('param_get_projection:&
-                   & too many projections defined')
+              if(counter+sites*sum(ang_states) > num_proj) &
+                 call io_error('param_get_projection: too many projections defined')
            end if
            !
            if(sites==-1) then
@@ -3093,8 +3094,8 @@
 
         ! check there are enough projections and add random projections if required
         if (.not. lpartrandom) then
-              if (counter.ne.num_proj) call io_error('param_get_projections:&
-                   & Fewer projections defined than the number of Wannier functions requested')
+              if (counter.ne.num_proj) call io_error(&
+               'param_get_projections: Fewer projections defined than the number of Wannier functions requested')
         else
            call random_seed()
            do loop=counter+1,num_proj
diff -urN wannier90-1.2.old/src/plot.F90 wannier90-1.2.new/src/plot.F90
--- wannier90-1.2.old/src/plot.F90	2009-10-29 12:15:29.000000000 +0100
+++ wannier90-1.2.new/src/plot.F90	2010-04-15 18:01:26.000000000 +0200
@@ -53,8 +53,8 @@
            if (all(kpt_latt(:,nkp)<eps6)) have_gamma=.true.       
        end do
        if(.not. have_gamma) &
-            write(stdout,'(1x,a)') '!!!! Kpoint grid does not include Gamma. &
-            &Interpolation may be incorrect. !!!!'
+            write(stdout,'(1x,a)') '!!!! Kpoint grid does not include Gamma. ',&
+            & ' Interpolation may be incorrect. !!!!'
        ! Transform Hamiltonian to WF basis
        !
        call hamiltonian_setup()
diff -urN wannier90-1.2.old/src/wannier_lib.F90 wannier90-1.2.new/src/wannier_lib.F90
--- wannier90-1.2.old/src/wannier_lib.F90	2010-01-24 17:50:12.000000000 +0100
+++ wannier90-1.2.new/src/wannier_lib.F90	2010-04-15 18:01:08.000000000 +0200
@@ -166,6 +166,12 @@
      exclude_bands_loc(1:num_exclude_bands) = exclude_bands(1:num_exclude_bands)
   end if
 
+  if (postproc_setup) then
+   call kmesh_write()
+   write(stdout,'(1x,a25,f11.3,a)') 'Time to write kmesh      ',io_time(),' (sec)'
+   write(stdout,'(/a)') ' '//trim(seedname)//'.nnkp written.'
+  endif
+
 
   call kmesh_dealloc()
   call param_dealloc()
